### PR TITLE
Using async in "max"

### PR DESCRIPTION
**Pull request by leny32**
Here is async compatibility with `max` from the issue: #125 

#### Checklist

- [ x ] run `npm run test` and `npm run benchmark`
- [ x ] tests and/or benchmarks are included
- [ ? ] documentation is changed or added
- [ x ] commit message and code follows the [Developer's Certification of Origin](https://github.com/fastify/.github/blob/master/CONTRIBUTING.md#developers-certificate-of-origin-11)
      and the [Code of conduct](https://github.com/fastify/.github/blob/master/CODE_OF_CONDUCT.md)

### CODE REVIEW

**Comment by leny32** [comment]
This worked when I last tested it.

File: index.js [hunk 1]
```diff
@@ -67,11 +67,11 @@ function rateLimitPlugin (fastify, settings, next) {
   }
 
   // onRoute add the onRequest rate-limit function if needed
-  fastify.addHook('onRoute', (routeOptions) => {
+  fastify.addHook('onRoute', async (routeOptions) => {
```
Comment (mcollina) [inline index.js:70]: Unfortunately using `async` here is not supported.

File: index.js [hunk 2]
```diff
@@ -67,11 +67,11 @@ function rateLimitPlugin (fastify, settings, next) {
...
     if (routeOptions.config && typeof routeOptions.config.rateLimit !== 'undefined') {
       if (typeof routeOptions.config.rateLimit === 'object') {
         const current = Object.create(pluginComponent)
-        const mergedRateLimitParams = makeParams(routeOptions.config.rateLimit)
+        const mergedRateLimitParams = await makeParams(routeOptions.config.rateLimit)
```
Comment (mcollina) [inline index.js:74]: why you need to do this asynchronously?

**Review by mcollina — CHANGES_REQUESTED** [review]
Can you please add a unit test?

File: index.js (hunk 1 above)
Comment (leny32) [reply index.js:70]: I'll fix this.

**Comment by leny32** [comment]
Has been tested using:
```js
const main = async (fastify) => {
	await fastify.register(require('fastify-rate-limit'), {
		keyGenerator: function (req) { return req.headers['authorization']; },
		max: async (_, key) => {
			let { access } = JSON.parse(await fastify.redis.get('API_KEY_' + key));
			return getAmount(access || 0);
		},
		skipOnError: true,
		errorResponseBuilder: function () {
			return {
				code: 429,
				error: 'Too Many Requests',
				message: 'Hourly limit reached, contact staff for increased limit',
				date: Date.now()
			};
		},
		timeWindow: 60 * 60 * 1000,
		redis: new ioredis({
			host: '127.0.0.1',
			port: 6379,
		})
	});
};
```

**Comment by mcollina** [comment]
Can you transform your snippet in a unit test? Check the `test/` folder.

**Comment by leny32** [comment]
I'm not very familiar with unit testing, but I'll try.

**Pull request merged**